timely_beliefs/beliefs/utils.py

```diff
@@ -1030,3 +1038,68 @@ def meta_repr(
             for attr in tb_structure._metadata
         ]
     )
+
+
+def upsample_beliefs_data_frame(
+    df: "classes.BeliefsDataFrame" | pd.DataFrame,
+    event_resolution: timedelta,
+    keep_nan_values: bool = False,
+) -> "classes.BeliefsDataFrame":
+    """Because simply doing df.resample().ffill() does not correctly resample the last event in the data frame.
+
+    :param df:                  In case of a regular pd.DataFrame, make sure to set df.event_resolution before passing it to this function.
+    :param event_resolution:    Resolution to upsample to.
+    :param keep_nan_values:     If True, place back resampled NaN values. Drops NaN values by default.
+    """
+    from_event_resolution = df.event_resolution
+    resample_ratio = pd.to_timedelta(to_offset(from_event_resolution)) / pd.Timedelta(
+        event_resolution
```

Comment:
What if `event_resolution == 0` (which I'm not sure what would mean in this context)?

Also, I'm thinking on the case of `from_event_resolution` is 0. The resample ratio would be 0 regardless of the value of `event_resolution`.

Reply:
The case of `event_resolution == timedelta(0)` is handled elsewhere, such that this method would not be called, but I'll add a `raise` statement regardless. The case of `from_event_resolution == timedelta(0)` is already handled by forward filling the instantaneous beliefs. I'm adding a test, which also lead me to a fix (for applying this function on BeliefsDataFrames rather than regular Pandas DataFrames).